### PR TITLE
Get products by slug api endpoint

### PR DESCRIPTION
**Pull request by paullla**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | mentioned in https://github.com/Sylius/Sylius/issues/12453
| License         | MIT

### CODE REVIEW

**Review by Roshyo — APPROVED** [review]

File: features/product/viewing_products/viewing_product_via_slug.feature [hunk 1]
```diff
@@ -0,0 +1,15 @@
+@viewing_products
+Feature: Viewing a product details using slug
+    In order to see products detailed information
+    As a Visitor
+    I want to be able to view a single product using slug
+
+    Background:
+        Given the store operates on a single channel in "United States"
+        And the store has a product "T-shirt banana"
+        And the description of product "T-shirt banana" is "You must have this beautiful T-shirt"
```
Comment (lchrusciel) [inline features/product/viewing_products/viewing_product_via_slug.feature:10]: not needed
```suggestion
```

File: features/product/viewing_products/viewing_product_via_slug.feature [hunk 2]
```diff
@@ -0,0 +1,15 @@
+@viewing_products
+Feature: Viewing a product details using slug
```
Comment (lchrusciel) [inline features/product/viewing_products/viewing_product_via_slug.feature:2]: Let's put it into https://github.com/Sylius/Sylius/blob/master/features/product/viewing_products/accessing_product_page_via_permalink.feature file, just below the already existing scenario. Also, can you add `no-api` tag to the scenario that is already there?

File: src/Sylius/Behat/Context/Api/Shop/ProductContext.php [hunk 3]
```diff
@@ -76,6 +76,29 @@ public function iOpenProductPage(ProductInterface $product): void
...
     }
 
+    /**
+     * @When I view product :product using slug
+     */
+    public function iViewProductUsingSlug(ProductInterface $product): void
+    {
+        /** @var ProductVariantInterface $productVariant */
+        $productVariant = $product->getVariants()->first();
+
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Api/Shop/ProductContext.php:86]:
```suggestion
```
is it needed?

File: src/Sylius/Behat/Context/Api/Shop/ProductContext.php [hunk 4]
```diff
@@ -76,6 +76,29 @@ public function iOpenProductPage(ProductInterface $product): void
...
+        $this->client->showByIri('/api/v2/shop/products/slug/'.$product->getSlug());
+
+        $this->sharedStorage->set('product', $product);
+        $this->sharedStorage->set('product_variant', $productVariant);
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Api/Shop/ProductContext.php:90]:
```suggestion
```

File: src/Sylius/Behat/Context/Api/Shop/ProductContext.php [hunk 5]
```diff
@@ -76,6 +76,29 @@ public function iOpenProductPage(ProductInterface $product): void
...
+    }
+
+    /**
+     * @Then I should be redirected to :product product
+     */
+    public function iShouldBeRedirectedToProduct(ProductInterface $product) {
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Api/Shop/ProductContext.php:96]: We are quite strongly following PSR coding style: https://www.php-fig.org/psr/psr-2/, therefore brackets have to be always on the next line. Also, we are adding return types everywhere
```suggestion
    public function iShouldBeRedirectedToProduct(ProductInterface $product): void
     {
```

File: src/Sylius/Bundle/ApiBundle/Controller/GetProductBySlugAction.php [hunk 6]
```diff
@@ -0,0 +1,63 @@
+<?php
+/*
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Controller/GetProductBySlugAction.php:2]:
```suggestion

/*
```

File: src/Sylius/Bundle/ApiBundle/Controller/GetProductBySlugAction.php [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+final class GetProductBySlugAction
+{
+    private ChannelContextInterface $channelContext;
+    private LocaleContextInterface $localeContext;
+    private ProductRepositoryInterface $productRepository;
+    private IriConverterInterface $iriConverter;
+    private RequestStack $requestStack;
+
+    public function __construct(ChannelContextInterface $channelContext, LocaleContextInterface $localeContext, ProductRepositoryInterface $productRepository, IriConverterInterface $iriConverter, RequestStack $requestStack)
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Controller/GetProductBySlugAction.php:32]:
```suggestion
    public function __construct(
        ChannelContextInterface $channelContext, 
        LocaleContextInterface $localeContext, 
        ProductRepositoryInterface $productRepository, 
        IriConverterInterface $iriConverter, 
        RequestStack $requestStack
    ) {
```

File: src/Sylius/Bundle/ApiBundle/Controller/GetProductBySlugAction.php [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+use ApiPlatform\Core\Api\IriConverterInterface;
+use Sylius\Component\Channel\Context\ChannelContextInterface;
+use Sylius\Component\Core\Repository\ProductRepositoryInterface;
+use Sylius\Component\Locale\Context\LocaleContextInterface;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+
+final class GetProductBySlugAction
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Controller/GetProductBySlugAction.php:24]: Let's write at least one contract test for this behaviour to check existence of query parameters

File: tests/Api/Shop/ProductsTest.php [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Tests\Api\Shop;
+
+use Sylius\Tests\Api\JsonApiTestCase;
+use Symfony\Component\HttpFoundation\Response;
+
+class ProductsTest extends JsonApiTestCase
```
Comment (lchrusciel) [inline tests/Api/Shop/ProductsTest.php:19]:
```suggestion
final class ProductsTest extends JsonApiTestCase
```

File: src/Sylius/Bundle/ApiBundle/Swagger/ProductSlugDocumentationNormalizer.php [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+    private const PRODUCT_SLUG_PATH = '/api/v2/shop/products/slug/{slug}';
+
+    private NormalizerInterface $decoratedNormalizer;
+
+    public function __construct(NormalizerInterface $decoratedNormalizer)
+    {
+        $this->decoratedNormalizer = $decoratedNormalizer;
+    }
+
+    public function supportsNormalization($data, $format = null)
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Swagger/ProductSlugDocumentationNormalizer.php:30]:
```suggestion
    public function supportsNormalization($data, $format = null): bool
```

File: src/Sylius/Behat/Context/Api/Shop/ProductContext.php [hunk 11]
```diff
@@ -76,6 +76,26 @@ public function iOpenProductPage(ProductInterface $product): void
...
+    }
+
+    /**
+     * @Then I should be redirected to :product product
+     */
+    public function iShouldBeRedirectedToProduct(ProductInterface $product): void
+    {
+        $response = $this->client->getLastResponse();
+
+        Assert::eq($response->headers->get('Location'), '/api/v2/shop/products/'.$product->getCode());
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Api/Shop/ProductContext.php:96]: Maybe we could use router to create this url?

**Review by Zales0123 — APPROVED** [review]

File: src/Sylius/Bundle/ApiBundle/Swagger/ProductSlugDocumentationNormalizer.php [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    public function supportsNormalization($data, $format = null): bool
+    {
+        return $this->decoratedNormalizer->supportsNormalization($data, $format);
+    }
+
+    public function normalize($object, $format = null, array $context = []): array
+    {
+        $docs = $this->decoratedNormalizer->normalize($object, $format, $context);
```
Comment (Zales0123) [inline src/Sylius/Bundle/ApiBundle/Swagger/ProductSlugDocumentationNormalizer.php:37]: I think that we need to cast the `normalize` function result to `(array)` to make the static analysis work :vulcan_salute:

**Review by lchrusciel — APPROVED** [review]

**Comment by lchrusciel** [comment]
Thanks, Paula! :tada:

**Pull request merged**